From bced6e2e595cd3b3df7a96bfbad6e0b8821550a9 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 27 Mar 2025 15:50:51 +0100
Subject: [PATCH 157/985] net: dovetail: restrict napi_complete_oob() to inband
 polling completion

napi_complete_oob() is called by napi_complete_done() to start
processing the ingress packets received from a device which is not
oob-capable. Make sure to restrict this call to the latter case,
i.e. when napi_complete_done() is running inband exclusively.

The out-of-band net stack is expected to schedule its RX thread
immediately on the oob receive path.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 net/core/dev.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 116f98b73..b52a4ac43 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -5292,6 +5292,12 @@ bool netif_receive_oob(struct sk_buff *skb)
 	 * packet is not picked, then decide whether we should leave
 	 * it to the in-band caller, or defer it to the RX softirq if
 	 * currently running oob.
+	 *
+	 * CAUTION: we _need_ the sirq indirection via inband_rx_work,
+	 * to make sure softirqs are going to be checked on return
+	 * from it. i.e. we could not simply set the RX softirq as
+	 * pending and be done with it, because we could not tell
+	 * whether it is going to be handled asap.
 	 */
 	if (dev && netif_oob_diversion(dev)) {
 		if (netif_deliver_oob(skb))
@@ -6478,9 +6484,9 @@ bool napi_complete_done(struct napi_struct *n, int work_done)
 	bool ret = true;
 
 	if (netif_oob_diversion(n->dev)) {
-		napi_complete_oob(n);
 		if (net_running_oob())
 			return true;
+		napi_complete_oob(n);
 	}
 
 	/*
-- 
2.30.2

